### PR TITLE
`Keccak` gates for zkVM

### PR DESCRIPTION
**Pull request by querolita**
This PR implements the Keccak gate as suggested in the [RFC](https://github.com/o1-labs/rfcs/pull/27).

It contains the constraints and layout of the initial state XOR and the permutation round itself. The witness generation code will come in a subsequent PR.

Closes 
https://github.com/o1-labs/proof-systems/issues/1214 
https://github.com/o1-labs/proof-systems/issues/1215 
https://github.com/o1-labs/proof-systems/issues/1216 
https://github.com/o1-labs/proof-systems/issues/1217 
https://github.com/o1-labs/proof-systems/issues/1222 
https://github.com/o1-labs/proof-systems/issues/1225

### CODE REVIEW

File: kimchi/src/circuits/argument.rs [hunk 1]
```diff
@@ -81,11 +81,38 @@ impl<F: Field, T: ExprOps<F>> ArgumentEnv<F, T> {
...
+    /// Witness cells in current row in an interval [from, to)
+    pub fn witness_curr_chunk(&self, from: usize, to: usize) -> Vec<T> {
+        let mut chunk = Vec::with_capacity(to - from);
+        for i in from..to {
+            chunk.push(self.witness_curr(i));
+        }
+        chunk
+    }
+
+    /// Witness cells in current row in an interval [from, to)
```
Comment (dannywillems) [inline kimchi/src/circuits/argument.rs:93]:
```suggestion
    /// Witness cells in next row in an interval [from, to)
```
?

File: kimchi/src/circuits/polynomials/keccak.rs [hunk 2]
```diff
@@ -21,65 +77,265 @@ use crate::circuits::{
...
+//~ | Columns  | [1540...1940) | [1940...2340) | [2340...2344 |
+//~ | -------- | ------------- | ------------- | ------------ |
+//~ | chi      | shifts_b      | shifts_sum    | f00          |
+//~
+//~ | Columns  | [0...4) | [4...100) |
+//~ | -------- | ------- | --------- |
+//~ | iota     | g00     | rest_g    |
+//~
+#[derive(Default)]
+pub struct KeccakRound<F>(PhantomData<F>);
```
Comment (dannywillems) [inline kimchi/src/circuits/polynomials/keccak.rs:144]: It is a single round of the permutation? (i.e. it is going to be called 24 times?)

File: kimchi/src/circuits/polynomials/keccak.rs [hunk 3]
```diff
@@ -21,65 +77,265 @@ use crate::circuits::{
...
+pub struct KeccakSponge<F>(PhantomData<F>);
+
+impl<F> Argument<F> for KeccakSponge<F>
+where
+    F: PrimeField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::KeccakSponge);
+    const CONSTRAINTS: u32 = 568;
+
+    // Constraints for one round of the Keccak permutation function
```
Comment (dannywillems) [inline kimchi/src/circuits/polynomials/keccak.rs:290]: Doesn't seem right.

File: kimchi/src/circuits/polynomials/keccak.rs [hunk 4]
```diff
@@ -21,65 +77,265 @@ use crate::circuits::{
...
-                if rot == 0 {
-                    continue;
+        // DEFINE ROUND CONSTANT
+        let rc = [env.coeff(0), env.coeff(1), env.coeff(2), env.coeff(3)];
+
+        // LOAD STATES FROM WITNESS LAYOUT
+        // THETA
+        let state_a = grid!(100, env.witness_curr_chunk(0, 100));
+        let state_c = grid!(20, env.witness_curr_chunk(100, 120));
+        let shifts_c = grid!(80, env.witness_curr_chunk(120, 200));
```
Comment (dannywillems) [inline kimchi/src/circuits/polynomials/keccak.rs:164]: called `reset_c` in the rfc

File: kimchi/src/circuits/polynomials/keccak.rs [hunk 5]
```diff
@@ -21,65 +77,265 @@ use crate::circuits::{
...
+        let dense_c = grid!(20, env.witness_curr_chunk(200, 220));
+        let quotient_c = grid!(20, env.witness_curr_chunk(220, 240));
+        let remainder_c = grid!(20, env.witness_curr_chunk(240, 260));
+        let bound_c = grid!(20, env.witness_curr_chunk(260, 280));
+        let dense_rot_c = grid!(20, env.witness_curr_chunk(280, 300));
+        let expand_rot_c = grid!(20, env.witness_curr_chunk(300, 320));
+        let state_d = grid!(20, env.witness_curr_chunk(320, 340));
+        let state_e = grid!(100, env.witness_curr_chunk(340, 440));
+        // PI-RHO
+        let shifts_e = grid!(400, env.witness_curr_chunk(440, 840));
```
Comment (dannywillems) [inline kimchi/src/circuits/polynomials/keccak.rs:174]: called reset_e in RFC

File: kimchi/src/circuits/polynomials/keccak.rs (hunk 3 above)
Comment (querolita) [reply kimchi/src/circuits/polynomials/keccak.rs:290]: Ahhh you're right, classical copy-paste typo

File: kimchi/src/circuits/polynomials/keccak.rs (hunk 4 above)
Comment (querolita) [reply kimchi/src/circuits/polynomials/keccak.rs:164]: reset is the canonical representation, shifts is for the "masks" shifted. I thought that was updated in the RFC as well? Do you have pointers?

File: kimchi/src/circuits/polynomials/keccak.rs (hunk 5 above)
Comment (querolita) [reply kimchi/src/circuits/polynomials/keccak.rs:174]: Same as above

File: kimchi/src/circuits/polynomials/keccak.rs (hunk 4 above)
Comment (dannywillems) [reply kimchi/src/circuits/polynomials/keccak.rs:164]: https://github.com/o1-labs/rfcs/blob/dc2458ab708e9a380628ecf0e77fb30157cad7e0/0010-zkvm-keccak.md

Comment (dannywillems) [reply kimchi/src/circuits/polynomials/keccak.rs:164]: https://github.com/o1-labs/rfcs/blob/dc2458ab708e9a380628ecf0e77fb30157cad7e0/0010-zkvm-keccak.md#step-theta

**Review by dannywillems — APPROVED** [review]

**Pull request merged**